docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+Internal Verona ABI (Draft)
+===========================
+
+Verona expects to use foreign-language ABIs (C/C++) for interoperability and reserves the right to change the internal ABI for Verona components between compiler releases.
+That said, it is useful for people working on the compiler, producing additional tooling, and writing tests for the ABI used internally to be documented.
+This draft is a proposal for the v1 ABI.
+
+Primitive types
+---------------
+
+The primitive types in Verona are:
+
+ - Boolean
+ - Integers, signed or unsigned, with 1-128 bits of precision
+ - Floating point values, with various power-of-two sizes
+ - Pointers to objects
+
+The representation of these may change when used in a complex type expression but in the absence of other constraints they will be stored as simple values, as described in the following sections.
+The compiler is free to relax any of these constraints for on-stack variables that are not passed by reference (for example, leaving some bits in a value undefined if the compiler can prove that they are never read).
+
+Verona places the following restrictions on target architectures:
+
+ - The size of a byte is 8 bits.
+ - The architecture can load and store a single byte without read-modify-write operations
+ - The order of bits within different-sized values is consistent
+ - Signed arithmetic on integers uses twos complement
+
+This is a slightly tighter set of requirements than the C specification but is looser than the de-facto C standard as used in large deployed codebases.
+
+### Booleans
+
+Boolean values are stored as bytes, with the bit pattern 0 or 1 in the platform's native endian.
+The compiler is responsible for ensuring that no bits other than the least-significant bit in a boolean are ever non-zero.
+
+The alignment of booleans is 1 byte.
+
+### Integers
+
+Power-of-two-sized integers in Verona can be used for arithmetic and so are expected to use the representation that the target architecture uses for integer operations.
+The types `U8`, `U16`, `U32`, `U64`, and `U128` are all unsigned fixed-width integers with the platform's native endian.
+The types `S8`, `S16`, `S32`, `S64`, and `S128` are all signed fixed-width integers represented as twos complement numbers with the platform's native endian.
+These types are always stored naturally aligned.
+For example, a `U64` is always stored on an 8-byte boundary.
+
+Other integer types, for example `U56` are *storage-only* types.
```

Comment:
I think there's a consensus building that non-power-of-two integers be left out of the ABI, at least for v1.

I'm not sure what kind of document you want to write: A strict one with only agreed things, or a more open one, with ideas for the future.

If the former, this paragraph should probably be removed, for now. If the latter, I'd rewrite this as "There's an open discussion about allowing non-power-of-two integers, in which case the semantic is...".

Reply:
I initially wrote this as a straw-man for discussion.  There are a few places where non-power-of-two integers (or, really, bitfields) can be really useful.  The two that come to mind are:

 - Implementing a generic Integer type that stores the value inline if it fits or as a pointer if it doesn't.  
 - String representations that want to store as many characters as possible in a pointer or a pointer to the real data for larger strings.

Reply:
Bitfields are an interesting case per se, but they can also be implemented in Verona over a larger type + shift and mask semantics, so they match as an `U32` for example.

The "generic integer" sounds to me like `(U8 | U16 | U32 | U64 | U128)` and therefore implemented by the standard ABI.

The string representation is an odd one, because only very short strings will be in this representation, so you get two very different performance characteristics with the same type. Not to mention if the string isn't plain ASCII, getting the actual byte representation might be non-trivial. I'd prefer to implement that in Verona...